apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/PackageNameProvider.kt

```diff
@@ -3,35 +3,75 @@ package com.apollographql.apollo.compiler
 import java.io.File
 
 class PackageNameProvider(
-    /**
-     * the root package name for all classes:
-     *
-     * fragments will use rootPackageName.fragment
-     * types will use rootPackageName.type
-     * operations, will use rootPackageName.{relativePosition to rootDir}
-     */
-    private val rootPackageName: String?,
-    private val rootDir: File?,
+    private val rootPackageName: String,
+
     /**
      * the package name for fragments and types will be written there
      */
-    val irPackageName: String,
+    val schemaPackageName: String,
+
     /**
      * the package name for fragments, types, queries and mutations
      * This will flatten all the classes in the same package name. If you have operations in subfolders,
      * use rootPackageName instead
      */
+    @Deprecated("")
     private val outputPackageName: String?
 ) {
-  fun fragmentsPackageName(): String {
-    return if (irPackageName.isNotEmpty()) "$irPackageName.fragment" else "fragment"
+  val fragmentsPackageName = outputPackageName?.appendPackageName("fragment") ?: rootPackageName
+      .appendPackageName(schemaPackageName)
+      .appendPackageName("fragment")
+
+  val typesPackageName = outputPackageName?.appendPackageName("type") ?: rootPackageName
+      .appendPackageName(schemaPackageName)
+      .appendPackageName("type")
+
+  fun operationPackageName(filePath: String): String {
+    if (outputPackageName != null) {
+      return outputPackageName
+    }
+
+    val packageName = filePath.formatPackageName(dropLast = 1)
+    if (packageName == null) {
+      throw IllegalArgumentException("graphql file must be placed under src/{foo}/graphql:\n$filePath")
+    }
+    return rootPackageName.appendPackageName(packageName)
   }
+}
 
-  fun typesPackageName(): String {
-    return if (irPackageName.isNotEmpty()) "$irPackageName.type" else "type"
+fun String.appendPackageName(packageName: String): String {
+  if (packageName.isBlank()) {
+    return this
   }
 
-  fun operationPackageName(filePath: String): String {
-   return outputPackageName ?: filePath.formatPackageName()
+  return if (this.isNotBlank()) {
+    "$this.$packageName"
+  } else {
+    packageName
   }
-}
\ No newline at end of file
+}
+
+fun File.child(vararg path: String): File {
+  return path.toList().fold(this) { file, p ->
+    File(file, p)
+  }
```

Comment:
This is meant to be used like this: `buildDir.child("generated", "source", "apollo", variantName, serviceVariant.name)`. There can be any number of subdirectories and this takes care of using the correct separator automatically.

Reply:
We could certainly join on `File.separator` too and then call File(parent, child) only once

Reply:
You could also use `File.resolve(String)` (stdlib extension function). I don't think it causes any issues with file separators.
On Windows:
```
File("C:\\Users\\steve\\").resolve("hello/test")
res0: java.io.File = C:\Users\steve\hello\test
```

Reply:
Looking at the [stdlib source](https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/jvm/src/kotlin/io/files/Utils.kt#L404), there doesn't seem to be anything specific for separators handling there. `File.resolve` is basically doing `File(baseName + File.separatorChar + relative`. Could it be that `File` itself is doing the magic ?

Reply:
According to [stackoverflow](https://stackoverflow.com/a/2417546/752630), using `'/'` should work on Windows as well without the need to actually use `File.separator`. There is some logic in the JDK code to [normalize](https://github.com/openjdk/jdk/blob/master/src/java.base/windows/classes/java/io/WinNTFileSystem.java#L87) paths. Although I'm not sure what this exactly means...

Reply:
`File.separatorChar` should be platform dependent, means it will be resolved to `/` on Unix, `\` on Windows.
So `File(file, path.toList().joinToString(separator = File.separatorChar))` should work fine

Reply:
Ok, I've done that. I'll try to build on a Windows machine one of these days. Maybe we don't even need to bother and just using '/' everywhere would do.